yksgi/src/lib.rs

```diff
@@ -313,8 +316,68 @@ impl SGInterp {
             // FIXME: When we see the control point we are done and can just return.
             return true;
         } else {
-            // FIXME: Properly implement calls.
-            todo!()
+            // We are going to do a native call via libffi, starting with looking up the address of
+            // the callee in the virtual address space.
+            //
+            // OPT: This could benefit from caching, and we may already know the address of the
+            // callee if we inlined it when preparing traces:
+            // https://github.com/ykjit/yk/issues/544
+            assert!(!LLVMIsAFunction(func).is_null());
+            let fptr = dlsym(ptr::null_mut(), name.as_ptr() as *const i8);
+            if fptr == ptr::null_mut() {
+                todo!("couldn't find symbol: {}", name);
+            }
+
+            if LLVMIsFunctionVarArg(LLVMGetElementType(LLVMTypeOf(func))) != 0 {
+                todo!("calling a varargs function");
+            }
+
+            // Now build the calling interface and collect the values of the callee's arguments.
+            //
+            // It may be tempting to use `LLVMGetArgOperand()` to extract callee operands from our
+            // `CallInst`, but at the time of writing this wraps the C++ method
+            // `FuncletPadInst::getArgOperand()` which is not what we want.
+            //
+            // `CallBase::getArgOperand()` isn't currently exposed by the LLVM C API, so we instead
+            // use the implicit knowledge that for a `CallInst` the first `N` operands are the
+            // operands of the function being called (the remaining operand is the callee).
```

Comment:
This is fine and there are various other places where we similarly need to use implicit knowledge about the order of arguments of a specific instruction. I don't see those special LLVM functions being exposed anytime soon as I imagine there's others.

Reply:
In light of that, is this comment too long?

Reply:
Yeah, I guess we could shorten it a bit. If we write long paragraphs for each weirdness/missing feature in the LLVM C API we'll be here for a long time.

Reply:
Aggressively shortened in 130766753 :)